File: ios/App/App/BackgroundGeoPlugin.swift
import Foundation
import Capacitor
import CoreLocation
import UIKit

/**
 * Aile Takibi - Arka plan GPS izleme eklentisi.
 * Core Location ile gerçek arka plan takibi sağlar ve
 * konum verilerini arka planda doğrudan Telegram'a gönderir.
 */
@objc(BackgroundGeoPlugin)
public class BackgroundGeoPlugin: CAPPlugin, CLLocationManagerDelegate, URLSessionDelegate {

    private let locationManager = CLLocationManager()
    private var isTracking = false
    private var pendingStartCall: CAPPluginCall?
    private var sendTimer: Timer?
    private var backgroundSession: URLSession?
    private var backgroundCompletionHandler: (() -> Void)?

    private var botToken = ""
    private var chatId = ""
    private var intervalMs = 5000
    private var lastSendAt: TimeInterval = 0

    override public func load() {
        // Background URLSession: delegate ile tutulur, arka planda tamamlanır
        let config = URLSessionConfiguration.background(withIdentifier: "com.irtibat.ailetakibi.telegram")
        config.sessionSendsLaunchEvents = true
        config.isDiscretionary = false
        config.timeoutIntervalForRequest = 30
        config.timeoutIntervalForResource = 60
        backgroundSession = URLSession(configuration: config, delegate: self, delegateQueue: nil)

        locationManager.delegate = self
        locationManager.desiredAccuracy = kCLLocationAccuracyBest
        locationManager.distanceFilter = kCLDistanceFilterNone
        locationManager.allowsBackgroundLocationUpdates = true
        locationManager.pausesLocationUpdatesAutomatically = false
        locationManager.showsBackgroundLocationIndicator = true

        UIDevice.current.isBatteryMonitoringEnabled = true
    }

    @objc func start(_ call: CAPPluginCall) {
        botToken = call.getString("botToken") ?? ""
        chatId = call.getString("chatId") ?? ""
        intervalMs = call.getInt("intervalMs", 5000) ?? 5000

        DispatchQueue.main.async {
            let status = CLLocationManager.authorizationStatus()
            switch status {
            case .authorizedAlways, .authorizedWhenInUse:
                self.beginTracking()
                call.resolve(["started": true])
            case .notDetermined:
                self.pendingStartCall = call
                self.locationManager.requestAlwaysAuthorization()
            case .denied, .restricted:
                call.reject("Konum izni reddedildi. Ayarlar'dan 'Her Zaman İzin Ver' seçin.")
            @unknown default:
                call.reject("Bilinmeyen izin durumu")
            }
        }
    }

    @objc func stop(_ call: CAPPluginCall) {
        isTracking = false
        sendTimer?.invalidate()
        sendTimer = nil
        locationManager.stopUpdatingLocation()
        call.resolve(["stopped": true])
    }

    @objc func getCurrentPosition(_ call: CAPPluginCall) {
        if let loc = locationManager.location {
            call.resolve([
                "lat": loc.coordinate.latitude,
                "lng": loc.coordinate.longitude,
                "accuracy": loc.horizontalAccuracy
            ])
        } else {
            call.reject("Konum henüz alınamadı")
        }
    }

    private func beginTracking() {
        isTracking = true
        lastSendAt = 0
        // Sürekli konum akışı: uygulamayı arka planda uyanık tutar
        locationManager.startUpdatingLocation()
        startSendTimer()
    }

    // Timer: konum güncellemesi gelmese bile son bilinen konumu gönderir
    private func startSendTimer() {
        sendTimer?.invalidate()
        let interval = max(Double(intervalMs) / 1000.0, 1.0)
        let timer = Timer(timeInterval: interval, repeats: true) { [weak self] _ in
            guard let self = self, self.isTracking else { return }
            if let loc = self.locationManager.location {
                self.sendLocation(loc)
            }
        }
        RunLoop.main.add(timer, forMode: .common)
        sendTimer = timer
    }

    // MARK: - CLLocationManagerDelegate

    public func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        guard let loc = locations.last else { return }
        let timestamp = Int(loc.timestamp.timeIntervalSince1970 * 1000)
        let payload: PluginCallResultData = [
            "lat": loc.coordinate.latitude,
            "lng": loc.coordinate.longitude,
            "accuracy": loc.horizontalAccuracy,
            "timestamp": timestamp
        ]
        notifyListeners("location", data: payload)
        sendLocation(loc)
    }

    public func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        let nsError = error as NSError
        if nsError.code == CLError.locationUnknown.rawValue {
            return
        }
        notifyListeners("error", data: ["message": error.localizedDescription])
    }

    public func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
        let status = CLLocationManager.authorizationStatus()
        if status == .authorizedAlways || status == .authorizedWhenInUse {
            if let call = pendingStartCall {
                pendingStartCall = nil
                beginTracking()
                call.resolve(["started": true])
            }
        } else if status == .denied {
            if let call = pendingStartCall {
                pendingStartCall = nil
                call.reject("Konum izni reddedildi")
            }
        }
    }

    private func sendLocation(_ loc: CLLocation) {
        guard isTracking else { return }
        let now = Date().timeIntervalSince1970
        let interval = Double(intervalMs) / 1000.0
        guard (now - lastSendAt) >= interval else { return }
        lastSendAt = now
        let timestamp = Int(loc.timestamp.timeIntervalSince1970 * 1000)
        sendToTelegram(lat: loc.coordinate.latitude, lng: loc.coordinate.longitude, timestamp: timestamp)
    }

    // MARK: - Telegram gönderimi (native, arka planda çalışır)

    private func batteryInfo() -> (level: Int, charging: Bool) {
        let level = Int(UIDevice.current.batteryLevel * 100)
        let state = UIDevice.current.batteryState
        return (level == -1 ? 100 : level, state == .charging || state == .full)
    }

    private func sendToTelegram(lat: Double, lng: Double, timestamp: Int) {
        guard !botToken.isEmpty, !chatId.isEmpty, let session = backgroundSession else { return }

        let battery = batteryInfo()
        let dateStr = dateString(from: timestamp)
        let text =
            "📍 Aile Takibi\n" +
            "⏰ \(dateStr)\n" +
            String(format: "🛰 Konum: %.6f, %.6f\n", lat, lng) +
            "🔗 Google Maps: https://www.google.com/maps?q=\(lat),\(lng)\n" +
            "📶 Bağlantı: bilinmiyor\n" +
            "🔋 Pil: %\(battery.level)\n" +
            "⚡ Şarj: \(battery.charging ? "Evet" : "Hayır")"

        guard let url = URL(string: "https://api.telegram.org/bot\(botToken)/sendMessage") else { return }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        let body: [String: Any] = [
            "chat_id": chatId,
            "text": text,
            "disable_web_page_preview": true
        ]
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)

        let task = session.dataTask(with: request)
        task.resume()
    }

    private func dateString(from timestampMs: Int) -> String {
        let date = Date(timeIntervalSince1970: TimeInterval(timestampMs) / 1000)
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "tr_TR")
        formatter.dateFormat = "dd.MM.yyyy HH:mm:ss"
        return formatter.string(from: date)
    }

    // MARK: - URLSessionDelegate (background session tamamlanması)

    public func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
        // Hata durumunda sessizce geç; bir sonraki aralıkta yeni gönderim dener
        if let error = error {
            NSLog("AileTakibi Telegram gönderim hatası: %@", error.localizedDescription)
        }
    }

    public func urlSessionDidFinishEvents(forBackgroundURLSession session: URLSession) {
        DispatchQueue.main.async {
            self.backgroundCompletionHandler?()
            self.backgroundCompletionHandler = nil
        }
    }
}

File: android/app/src/main/java/com/irtibat/ailetakibi/BackgroundGeoPlugin.java
package com.irtibat.ailetakibi;

import android.Manifest;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.PackageManager;
import android.os.BatteryManager;
import android.os.Build;
import android.location.Location;

import androidx.core.content.ContextCompat;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import com.getcapacitor.annotation.PermissionCallback;
import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationServices;

import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

/**
 * Aile Takibi - Arka plan GPS izleme eklentisi (Android).
 * Foreground Service üzerinden konum alır ve arka planda doğrudan Telegram'a gönderir.
 */
@CapacitorPlugin(
    name = "BackgroundGeo",
    permissions = {
        @Permission(alias = "location", strings = {
            Manifest.permission.ACCESS_FINE_LOCATION,
            Manifest.permission.ACCESS_COARSE_LOCATION
        })
    }
)
public class BackgroundGeoPlugin extends Plugin {

    private LocationTracker tracker;
    private String botToken = "";
    private String chatId = "";
    private long intervalMs = 5000;
    private long lastSendAt = 0;

    @Override
    public void load() {
        tracker = new LocationTracker(getContext());
        tracker.setListener((location) -> {
            JSObject payload = new JSObject();
            payload.put("lat", location.getLatitude());
            payload.put("lng", location.getLongitude());
            payload.put("accuracy", location.getAccuracy());
            payload.put("timestamp", location.getTime());
            notifyListeners("location", payload);

            long now = System.currentTimeMillis();
            if (now - lastSendAt >= intervalMs) {
                lastSendAt = now;
                sendToTelegram(location.getLatitude(), location.getLongitude(), location.getTime());
            }
        });
    }

    @PluginMethod
    public void start(PluginCall call) {
        botToken = call.getString("botToken", "");
        chatId = call.getString("chatId", "");
        intervalMs = call.getInt("intervalMs", 5000);
        lastSendAt = 0;

        if (!hasLocationPermission()) {
            requestPermissionForAlias("location", call, "locationPermsCallback");
            return;
        }
        beginTracking(call);
    }

    @PermissionCallback
    private void locationPermsCallback(PluginCall call) {
        if (hasLocationPermission()) {
            beginTracking(call);
        } else {
            call.reject("Konum izni verilmedi");
        }
    }

    private void beginTracking(PluginCall call) {
        if (tracker != null) tracker.start();
        call.resolve();
    }

    @PluginMethod
    public void stop(PluginCall call) {
        if (tracker != null) tracker.stop();
        call.resolve();
    }

    @PluginMethod
    public void getCurrentPosition(PluginCall call) {
        FusedLocationProviderClient client = LocationServices.getFusedLocationProviderClient(getContext());
        try {
            client.getLastLocation().addOnSuccessListener(loc -> {
                if (loc == null) {
                    call.reject("Konum henüz alınamadı");
                    return;
                }
                JSObject result = new JSObject();
                result.put("lat", loc.getLatitude());
                result.put("lng", loc.getLongitude());
                result.put("accuracy", loc.getAccuracy());
                result.put("timestamp", loc.getTime());
                call.resolve(result);
            }).addOnFailureListener(e -> call.reject("Konum alınamadı: " + e.getMessage()));
        } catch (SecurityException e) {
            call.reject("Konum izni yok");
        }
    }

    private boolean hasLocationPermission() {
        return ContextCompat.checkSelfPermission(getContext(), Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED
            || ContextCompat.checkSelfPermission(getContext(), Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED;
    }

    // MARK: - Telegram gönderimi (native, arka planda çalışır)

    private int[] batteryInfo() {
        IntentFilter filter = new IntentFilter(Intent.ACTION_BATTERY_CHANGED);
        Intent status = getContext().registerReceiver(null, filter);
        int level = -1;
        int scale = -1;
        boolean charging = false;
        if (status != null) {
            level = status.getIntExtra(BatteryManager.EXTRA_LEVEL, -1);
            scale = status.getIntExtra(BatteryManager.EXTRA_SCALE, -1);
            int plugged = status.getIntExtra(BatteryManager.EXTRA_PLUGGED, 0);
            charging = plugged != 0;
        }
        int pct = (scale > 0) ? (int)((level * 100) / scale) : 100;
        return new int[]{ pct, charging ? 1 : 0 };
    }

    private void sendToTelegram(double lat, double lng, long timestampMs) {
        if (botToken.isEmpty() || chatId.isEmpty()) return;

        int[] battery = batteryInfo();
        SimpleDateFormat sdf = new SimpleDateFormat("dd.MM.yyyy HH:mm:ss", Locale.forLanguageTag("tr"));
        String dateStr = sdf.format(new Date(timestampMs));
        String text = "📍 Aile Takibi\n" +
            "⏰ " + dateStr + "\n" +
            String.format(Locale.US, "🛰 Konum: %.6f, %.6f\n", lat, lng) +
            "🔗 Google Maps: https://www.google.com/maps?q=" + lat + "," + lng + "\n" +
            "📶 Bağlantı: bilinmiyor\n" +
            "🔋 Pil: %" + battery[0] + "\n" +
            "⚡ Şarj: " + (battery[1] == 1 ? "Evet" : "Hayır");

        final String body = "{\"chat_id\":\"" + chatId + "\",\"text\":\"" + escapeJson(text) + "\",\"disable_web_page_preview\":true}";

        new Thread(() -> {
            try {
                URL url = new URL("https://api.telegram.org/bot" + botToken + "/sendMessage");
                HttpURLConnection conn = (HttpURLConnection) url.openConnection();
                conn.setRequestMethod("POST");
                conn.setRequestProperty("Content-Type", "application/json");
                conn.setDoOutput(true);
                conn.setConnectTimeout(15000);
                conn.setReadTimeout(15000);
                OutputStream os = conn.getOutputStream();
                os.write(body.getBytes(StandardCharsets.UTF_8));
                os.flush();
                os.close();
                conn.getInputStream().close();
                conn.disconnect();
            } catch (Exception ignored) {}
        }).start();
    }

    private String escapeJson(String s) {
        return s.replace("\\", "\\\\").replace("\"", "\\\"").replace("\n", "\\n").replace("\r", "");
    }
}